v2/pkg/subscraping/sources/shodan/shodan.go

```diff
@@ -75,10 +76,13 @@ func (s *Source) Run(ctx context.Context, domain string, session *subscraping.Se
 			}
 
 			for _, data := range response.Subdomains {
-				results <- subscraping.Result{
-					Source: s.Name(), Type: subscraping.Subdomain, Value: fmt.Sprintf("%s.%s", data, domain),
+				value := fmt.Sprintf("%s.%s", data, response.Domain)
+				if strings.HasSuffix(value, domain) {
```

Comment:
@dwisiswant0 I realized, `HasSuffix` might not be the best, I think actually this could act like a `grep`.

For example, if user supplies `api.veracode.com`, this could match `stg-api.veracode.com` which is not strictly speaking, underneath `.api.veracode.com` ? 
Or is it ok to keep it fuzzy ?

Reply:
I was just about to review this.

It seems that we might not even need to use `strings` pkg here. Since the API actually returns both `domain`  `subdomains`, we could just concatenate them directly without any issues. This way, we avoid any perf hit that could come from relying on the `strings` package.

To address your concern; we can put that aside for now, but we should definitely take note of it (https://github.com/projectdiscovery/subfinder/issues/1389#issuecomment-2376145684). No matter what input we provide, Shodan will return the `subdomains` associated with the root `domain`.

Reply:
Ok great, is there filtering done later in the post-processing of all sources in subfinder ?
I would assume if the user gives a subdomain, they only expect children under that ?

Reply:
Wait, I changed my mind.

Reply:
@dogancanbakir  - at first I also thought that we should align the behavior with the API provider. But it seems reasonable to apply a special case for Shodan API response since the output may not meet expectations without specific treatment to filter it.

Reply:
From my other comment with the screenshot in the debugger, I saw that [there is indeed a filter ](https://github.com/projectdiscovery/subfinder/blob/6537b327867861ec4eb90d07604f8bb79bdce88d/v2/pkg/runner/enumerate.go#L67)at the end of the whole pipeline for collecting all sources, so maybe my modification for shodan is redundant, we could just concatenate parent domain + subdomains and return all subdomains, and let the existing filter do the work, to avoid unnecessary `strings` usage in Shodan ?

Reply:
Oh, I didn't realize that either, my bad, folks.

I'll go ahead and revert the changes and stick to "just concatenate them directly" as I proposed.